main.c

```diff
@@ -3,106 +3,179 @@
+consumer. The consumer reports an error and kills the program if it detects that it has received messages
+out of order, which indicates a serious failure by the program. At the end, it prints how many messages
+it successfully read.
+
+The PRODUCER_SLOWDOWN and CONSUMER_SLOWDOWN allow you to make producers and consumers randomly lag during 
+operation, to simulate doing expensive operations. Slowing down the consumer leads to far fewer messages
+being recieved, as they are overwritten by the faster producer. Slowing down the producer causes all or 
+nearly all messages to get read by the consumer. The slowdown works by randomly choosing a number x in the 
+range [0, SLOWDOWN] and incrementing an integer from 0 to x in a while loop.
+(nb - it is a very low variance randomness. might be interesting to use an exponential distribution instead
+of uniform)
+
 */
 
-#define BUFFER_SIZE 4
+#define BUFFER_SIZE 1000
+#define FINAL_MESSAGE 500000
+#define PRODUCER_SLOWDOWN 200
+#define CONSUMER_SLOWDOWN 200
 
 struct message {
 	int content;
 	uint32_t locked; 
 };
 
-int finished=0;
+int die_immediately=0;
 struct message buffer[BUFFER_SIZE];
+pthread_t consumer_thread, producer_thread;
 
+void slowdown(int n) {
+	int target = (int) (drand48() * n);
+	int slow = 0;
+	while (slow++ < target) {}
+}
 
-void producer() {
-	int pos = 0;
+/*
+ * Prints the contents of the buffer to stdout.
+ * Construct a string and then print it rather than printing incrementally,
+ * so that the threads don't splatter each other.
+ */
+void printBuffer() {
+	char str[100] = "[";
+	char next[100];
+	int i;
+	for (i=0; i<BUFFER_SIZE; i++) {
+		struct message m = buffer[i];
+		if (m.locked == 1) {
+			strcat(str, "X,");
+		} else {
+			sprintf(next, "%d,", m.content % 100);
+			strcat(str, next);
+		}
+	}
+	strcat(str, "]\n");
+	printf("%s",str);
+}
+
+void *producer(void *_) {
+	int pos = BUFFER_SIZE - 1;
 	buffer[pos].locked = 1;
 	int i;
-	for (i=0; i<10; i++) {
+	for (i=0; i<=FINAL_MESSAGE; i++) {
+		slowdown(PRODUCER_SLOWDOWN);
 		int next = pos;
 		next = (next + 1) % BUFFER_SIZE;
-		// while (1) {
-		// 	if (!OSAtomicTestAndSet(1, &(buffer[next].locked))) {
-		// 		break;
-		// 	}
-		// }
-		printf("at i=%d, %d -> %d\n", next, buffer[next].content, i);
+		while (1) {
+			if (!OSAtomicTestAndSet(1, &(buffer[next].locked))) {
+				break;
+			}
+		}
+		if (die_immediately) {
+			return NULL;
+		}
 		buffer[next].content = i;
 		buffer[pos].locked = 0;
 		pos = next;
 	}
 	buffer[pos].locked = 0;
-	for (i=0; i<BUFFER_SIZE; i++) {
-		printf("\t%d:%d\n", i, buffer[i].content);
-	}
+
+	return NULL;
 }
 
-void printBuffer() {
-	int i;
-	printf("[");
-	for (i=0; i<BUFFER_SIZE; i++) {
-		struct message m = buffer[i];
-		if (m.locked == 1) {
-			printf("X,");
-		} else {
-			printf("%d,", m.content);
+void *consumer(void *_) {
+	int pos = 0;
+	int val;
+	int numRead = 0;
+	int lastVal = -1;
+	while (1) {
+		slowdown(CONSUMER_SLOWDOWN);
```

Comment:
shouldn't the slowdown happened while the slot is locked? presumably the slowdown is due to the cost of processing the message.